### PR TITLE
Add point in time to HLRC

### LINKED ISSUE #70593: Document how to use search_after w/ a point in time (PIT) via HLRC
In modern versions of Elasticsearch, the recommended approach for deep pagination is to use `search_after` [1] with a point in time (PIT) [2].

It'd be helpful to document how to perform this type of search/pagination with the HLRC.

[1] https://www.elastic.co/guide/en/elasticsearch/reference/7.11/paginate-search-results.html#search-after
[2] https://www.elastic.co/guide/en/elasticsearch/reference/7.11/point-in-time-api.html

### PR DESCRIPTION
**Pull request by dnhatn**
Closes #70593

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/ClearScrollResponse.java [hunk 1]
```diff
@@ -30,8 +30,8 @@
     private static final ParseField SUCCEEDED = new ParseField("succeeded");
     private static final ParseField NUMFREED = new ParseField("num_freed");
 
-    private static final ConstructingObjectParser<ClearScrollResponse, Void> PARSER = new ConstructingObjectParser<>("clear_scroll",
-            true, a -> new ClearScrollResponse((boolean)a[0], (int)a[1]));
+    private static final ConstructingObjectParser<ClosePointInTimeResponse, Void> PARSER = new ConstructingObjectParser<>("clear_scroll",
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/action/search/ClearScrollResponse.java:33]: Can you please explain what's the need for this change?

File: server/src/main/java/org/elasticsearch/action/search/OpenPointInTimeResponse.java [hunk 2]
```diff
@@ -12,40 +12,55 @@
...
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ObjectParser;
 import org.elasticsearch.common.xcontent.ToXContentObject;
 import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
 
 import java.io.IOException;
 import java.util.Objects;
 
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/action/search/OpenPointInTimeResponse.java:24]: It is good to add `OpenPointInTimeResponseTests` to check serialization from and to XContent. But I don't mind if we do it in a following PR.

File: docs/java-rest/high-level/search/point-in-time.asciidoc [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+A point in time must be opened before being used in search requests.
+An OpenPointInTimeRequest requires an `index` and `keepAlive` arguments:
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/SearchDocumentationIT.java[open-point-in-time]
+--------------------------------------------------
+<1> Create an `OpenPointInTimeRequest` with the target indices
+<2> Set the `keep_alive` parameter, which tells Elasticsearch how long it
```
Comment (mayya-sharipova) [inline docs/java-rest/high-level/search/point-in-time.asciidoc:13]: may be worth adding that this `keep_alive` is required parameter.

**Review by mayya-sharipova — APPROVED** [review]
@dnhatn Thanks, great work,  overall the PR LGMT!  I've left a couple of comments.  Also in the future we may want to have separate client classes for  `OpenPointInTimeRequest` and `OpenPointInTimeResponse` .

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 4]
```diff
@@ -1409,6 +1417,60 @@ public void testExplain() throws IOException {
...
+                String preference = randomAlphaOfLengthBetween(1, 10);
+                openRequest.preference(preference);
+                expectedParams.put("preference", preference);
+            }
+            openRequest.indicesOptions(setRandomIndicesOptions(openRequest.indicesOptions(), expectedParams));
+            final Request request = RequestConverters.openPointInTime(openRequest);
+            assertThat(request.getParameters(), equalTo(expectedParams));
+            assertThat(request.getMethod(), equalTo(HttpPost.METHOD_NAME));
+            final String expectedEndpoint;
+            if (indices.length == 0) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: Good catch. I fixed it in https://github.com/elastic/elasticsearch/pull/72167/commits/f0341ef1a13bf531b1d037225f4709c904c1e4de.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 5]
```diff
@@ -1914,9 +1976,15 @@ private static void setRandomSearchParams(SearchRequest searchRequest,
...
         }
         if (randomBoolean()) {
-            searchRequest.setCcsMinimizeRoundtrips(randomBoolean());
+            boolean ccsMinimizeRoundtrips = randomBoolean();
+            searchRequest.setCcsMinimizeRoundtrips(ccsMinimizeRoundtrips);
+            if (ccsMinimizeRoundtrips == false) {
+                expectedParams.put("ccs_minimize_roundtrips", "false");
+            }
+        }
+        if (searchRequest.isCcsMinimizeRoundtrips() == false) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: Yes, fixed in https://github.com/elastic/elasticsearch/pull/72167/commits/f0341ef1a13bf531b1d037225f4709c904c1e4de.

File: docs/java-rest/high-level/search/point-in-time.asciidoc [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+<1> Use the preference parameter e.g. to execute the search to prefer local
+shards. The default is to randomize across shards.
+
+=== Search with point in time
+A point in time can be passed to a search request via a PointInTimeBuilder,
+which requires a point in time ID returned from the open API.
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/SearchDocumentationIT.java[search-point-in-time
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply docs/java-rest/high-level/search/point-in-time.asciidoc]: Fixed in https://github.com/elastic/elasticsearch/pull/72167/commits/60f620125a4050200a6c2bdeb1b60b18e6c2014f :)

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java [hunk 7]
```diff
@@ -1283,6 +1287,66 @@ public final Cancellable clearScrollAsync(ClearScrollRequest clearScrollRequest,
...
+
+    /**
+     * Asynchronously open a point in time before using it in search requests
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/master/point-in-time-api.html"> Point in time API </a>
+     * @param openRequest the open request
+     * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
+     * @param listener the listener to be notified upon request completion
+     * @return a cancellable that may be used to cancel the request
+     */
+    public final Cancellable openPointInTimeAsync(OpenPointInTimeRequest openRequest,
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:1311]: Yes, it should be fast unless the request hits many indices or the cluster is not functioning. But here I merely follow the convention in HLRC that is returning Cancellable in *Async methods.

File: server/src/main/java/org/elasticsearch/action/search/ClearScrollResponse.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/ClearScrollResponse.java:33]: This is a bit messy. ClosePointInTimeResponse extends ClearScrollResponse, and this change just ensures that `fromXContent` method returns an instance of ClosePointInTimeResponse. I will work on a small PR to clean up this.

File: docs/java-rest/high-level/search/point-in-time.asciidoc (hunk 3 above)
Comment (dnhatn) [reply docs/java-rest/high-level/search/point-in-time.asciidoc]: ++, adjusted in https://github.com/elastic/elasticsearch/pull/72167/commits/754a2ed9e3050894a9c9d5a4d6a8aadfc4eaa6a9

File: server/src/main/java/org/elasticsearch/action/search/OpenPointInTimeResponse.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/OpenPointInTimeResponse.java:24]: Yes, please see https://github.com/elastic/elasticsearch/pull/72167#discussion_r631389330

**Comment by dnhatn** [comment]
@mayya-sharipova Thanks so much for your review.

**Comment by javanna** [comment]
@dnhatn I am not finding the newly added docs in the 7.x nor master docs for the high level REST client. Maybe I am looking in the wrong place. Can you double check that it got published and it is where you would expect it?

**Pull request merged**